### PR TITLE
1368: Add I18n to JavaScript

### LINKED ISSUE #1368: Missing i18n for validation error messages and JavaScript messages
- [x] The fields validation error messages are missing i18n. Addressed in PR #1370
- [x] JavaScript messages for glossary and translations are missing i18n. Addressed in PR #1369

### PR DESCRIPTION
**Pull request by pedro-mendonca**
Partially fixes (JS i18n) #1368

@ocean90 this is built on top of some of the changes of #1364 and assumes the renaming of the callbacks _or_ to _and_.
So is better to first merge #1364 and then fix any necessary naming.

~Feel free to add feedback on consolidate some more.~
~I just moved all the strings from JS to PHP and tried to sort it in a way it's easier for review.~

Edited: This PR doesn't use JS strings in PHP anymore. Just adds common `wp.i18n` to JavaScript.

### CODE REVIEW

File: assets/js/common.js [hunk 1]
```diff
@@ -43,7 +43,7 @@ var $gp = (
 				},
 			},
 			esc_html: function( s ) {
-				return $( '<div/>' ).text( s ).html();
+				return $( '</div>' ).text( s ).html();
```
Comment (ocean90) [inline assets/js/common.js]: `<div/>` is a shortcut for `<div></div>` so this should stay.

**Review by ocean90 — COMMENTED** [review]
Thanks for trying to improve I18N in GlotPress itself! But when making such large changes it's better to discuss them first so feedback can be provided earlier.

These are two separate issues which again should be handled in separate pull requests. For the JavaScript part it makes more sense to refactor this by using the wp-i18n JavaScript package and its `wp.i18n` API.

For the validation part, I'm not sure why you added the `$message` parameter to the second position. This seems like an unnecessary back-compat break.
Building the message like this is also still cumbersome and not easy to translate since the sentence is split into multiple parts.
I think it would be much easier if the full message, one for positive and one for negative can be provided and only the type and field names are (optionally) injected.

**Comment by pedro-mendonca** [comment]
Hi, I understand, and I agree.
About the JS, as the time is limited I tried to fix things following the same pattern as already exists assuming it won't be necessary to discuss as no code pattern was changed, making it easier for everyone. But I'll go for the new `wp.i18n` as it's a better solution.
About the consistency of strings listed above, if no suggestions, I'll keep em as they are.
Glossary and Translation -> verb **Save**
Priority and Status -> verb **Set**

For the validation error messages I can revert and open a separate PR.

**Comment by pedro-mendonca** [comment]
@ocean90 I've reduced this PR to only JS i18n.
Please let me know if it's ok now.

File: assets/js/common.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
-/* global document */
+/* global document, wp */
+const { __ } = wp.i18n;
```
Comment (ocean90) [inline assets/js/common.js]: Let's not use ES6 just yet. You could write this also as `var __ = wp.i18n.__`.

But we shouldn't make all the API methods available in the global window object. So do the extra variable only when necessary in the relevant method.

File: assets/js/common.js [hunk 3]
```diff
@@ -28,7 +29,7 @@ var $gp = (
 					$gp.notices.element.stop( true, true );
 
 					if ( true === dismissable ) {
-						dismissButton = ' <button type="button" class="button is-link gp-js-message-dismiss">' + $gp.l10n.dismiss + '</button>';
+						dismissButton = ' <button type="button" class="button is-link gp-js-message-dismiss">' + __( 'Dismiss', 'glotpress' ) + '</button>';
```
Comment (ocean90) [inline assets/js/common.js]:
```suggestion
						dismissButton = ' <button type="button" class="button is-link gp-js-message-dismiss">' + wp.i18n.__( 'Dismiss', 'glotpress' ) + '</button>';
```

No need for an extra variable.

File: assets/js/mass-create-sets-page.js [hunk 4]
```diff
@@ -37,15 +37,18 @@ jQuery( function( $ ) {
 					html += '</li>';
 					return html;
 				}
-				preview_html += preview_html_for( 'added', '{count} set(s) will be added' );
-				preview_html += preview_html_for( 'removed', '{count} set(s) will be removed' );
+				/* translators: {count}: Number of translation sets. */
```
Comment (ocean90) [inline assets/js/mass-create-sets-page.js]:
```suggestion
				/* translators: %s: Number of translation sets. */
```

File: assets/js/mass-create-sets-page.js [hunk 5]
```diff
@@ -37,15 +37,18 @@ jQuery( function( $ ) {
...
+				preview_html += preview_html_for( 'added', __( '%s set(s) will be added', 'glotpress' ) );
```
Comment (ocean90) [inline assets/js/mass-create-sets-page.js]: We can now make use of `_n()`.

**Review by ocean90 — DISMISSED** [review]
I think it's okay to use `wp.i18n.{methodName}` everywhere instead of having an extra variable.

File: assets/js/common.js (hunk 2 above)
Comment (pedro-mendonca) [reply assets/js/common.js]: Reverted back to ES5

File: assets/js/mass-create-sets-page.js (hunk 5 above)
Comment (pedro-mendonca) [reply assets/js/mass-create-sets-page.js]: Done with `_n()`

**Comment by pedro-mendonca** [comment]
As more and more JS is being added in functionalities like Local Translations branch, it would be good to make the current JS fully i18nized. Would you please recheck this PR @ocean90?
With the recent active dev team, should I also request review to @amieiro and @akirk?

**Review by amieiro — COMMENTED** [review]
To test this functionality, I have created the .pot file to add the new strings to the translation:

`wp i18n make-pot . languages/glotpress.pot`

This command adds the new strings:

image

Then I have imported this file in my local GlotPress and I have translated some strings like "Dismiss":

image

Then I have created the .mo file (glotpress-gl_ES.mo as I am testing with Galician language) and I have moved it to the `wp-content/language/plugins`, overwriting the current file.

I have reloaded the page content, to load the new translations, and then I have tried to translate a string with empty content, to fire an error, and the "Dismiss" text is not translated.  

image

The updates in the JavaScript file are loaded fine, but the translation in the JavaScript file is not showed.

I have updated the file version for the `common.js` file, using the `filemtime` function, to look for some cache problem, but the result is the same. 

`wp_register_script( 'gp-common', $url . '/common' . $suffix, array( 'jquery', 'wp-i18n' ), filemtime(plugin_dir_path( __FILE__ ) . '/../assets/js/common.js') );`


Am I missing some steps to test this code?

**Comment by pedro-mendonca** [comment]
Have you generated the .json files from the current .po with those new strings translated?
https://developer.wordpress.org/cli/commands/i18n/make-json/

**Comment by pedro-mendonca** [comment]
For JavaScript translations to be loaded you need those .json files generated from your `glotpress-gl_ES.po` on the same `./wp-content/languages/plugins/` folder.

File: assets/js/editor.js [hunk 6]
```diff
@@ -342,12 +344,26 @@ $gp.editor = (
...
+					case 'current':
+						status_name = wp.i18n._x( 'current', 'Single Status', 'glotpress' );
+						break;
+					case 'rejected':
+						status_name = wp.i18n._x( 'rejected', 'Single Status', 'glotpress' );
+						break;
+					case 'fuzzy':
+						status_name = wp.i18n._x( 'fuzzy', 'Single Status', 'glotpress' );
+						break;
+				}
```
Comment (amieiro) [inline assets/js/editor.js:366]: I think we need to add the "changes requested" status here:
```suggestion
					case 'changesrequested':
						status_name = wp.i18n._x( 'changes requested', 'Single Status', 'glotpress' );
						break;
				}
```

Comment (amieiro) [reply assets/js/editor.js:366]: Yes, you are right! I have added the .json files with the `wp i18n make-json glotpress-gl_ES.po` command in the `wp-content/languages/plugins` and the PR works fine!

Comment (pedro-mendonca) [reply assets/js/editor.js:366]: Indeed, will commit it right away.

**Review by amieiro — APPROVED** [review]

**Pull request merged**